### DISCUSSION TITLE
Performance?

### DISCUSSION THREAD

**Discussion by MrScautHD** [notgiven688/jitterphysics2#101] [How do I ... ?]
Hello, i really would love to use Jitter Physics but the performance is very bad... why is that so?

**Comment by MrScautHD** [comment]
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/6708993a-844e-49e5-a2a2-32503dca896a)

22 FPS and i have the latest ryzon processor...

**Comment by notgiven688** [comment]
Are you using the debug build? Use `dotnet run -c Release` to start in release build. This scene runs with 600 fps on my workstation from 2014.

**Reply by MrScautHD** [reply]
i see the release thing improves it but the performance is still very bad.

**Comment by MrScautHD** [comment]
it goves not even higher then 100 FPS

**Comment by notgiven688** [comment]
Multithreading activated?

edit: The checkbox "Multithreading"; which seems to be enabled in your screenshot.

**Reply by MrScautHD** [reply]
Relase mode
FPS: 32
Multithreading on
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/636dc4e3-318b-4aa8-90d5-4bc989100a78)

**Reply by notgiven688** [reply]
Interesting. And the pyramid scene from before? Which processor do you have?

**Reply by MrScautHD** [reply]
80 FPS when i trigger it

**Reply by MrScautHD** [reply]
Do you have Intel or AMD maybe is there a issue.

**Reply by MrScautHD** [reply]
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/d36ff282-d11b-4911-81da-e44b472ba60b)

**Reply by MrScautHD** [reply]
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/d159ef15-1306-4ceb-856d-861139c822e4)

**Reply by notgiven688** [reply]
Even single threaded I get ~100fps

![image](https://github.com/notgiven688/jitterphysics2/assets/37874600/556afdf1-c726-48e6-a866-f22b7f63c5b3)

Intel(R) Xeon(R) CPU E5-4650L 0 @ 2.60GHz; Linux

---

~200fps single-threaded on my notebook i7-1165G7; Windows

**Reply by MrScautHD** [reply]
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/c9a5350e-abab-40c3-8380-eb143b518e1d)

**Reply by MrScautHD** [reply]
Can you trigger it?

**Comment by MrScautHD** [comment]
Btw if you move them to fast with the mouse it crash:
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/2fa29e8b-b9a2-4e96-8d2a-9b2b1cf32377)

**Reply by notgiven688** [reply]
Looks like a different problem?

**Reply by MrScautHD** [reply]
yep

**Reply by notgiven688** [reply]
Limited the maximum pressure force in the demo scene:
https://github.com/notgiven688/jitterphysics2/pull/102

This should make the simulation more stable. However, please be aware that manipulating the sphere by grabbing and moving it at high velocities may still result in instability in the soft body simulation. In such cases, increasing the number of solver iterations is necessary.

**Comment by MrScautHD** [comment]
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/be72b49c-fdcf-4598-a7ee-5a5e253a7022)

**Comment by notgiven688** [comment]
AMD 5800 x3 results for the pyramid scene (edit: scene "Ancient Pyramids") are in:

244 fps single-threaded; 1401 fps multi-threaded

Something is odd.

**Reply by MrScautHD** [reply]
hu, thats wierd

**Reply by MrScautHD** [reply]
btw, how you got i managed to have more then 100 FPS?

**Reply by MrScautHD** [reply]
i set the target FPS to 1000 and im still low...
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/63dc948b-738c-43bb-8e20-ef1d13d0a824)

**Reply by notgiven688** [reply]
There is a frame rate counter for the physics system in your screenshot (below the histogram): 40 fps. Since the physics system is running quite slow on your system you only get 33 fps in total (graphics + physics, yellow number).

**Reply by MrScautHD** [reply]
ah ok, so do you know what the issue is?

**Reply by notgiven688** [reply]
I do not know what the issue is. You can try to manually set a thread-count (please use the "Ancient Pyramids" scene for comparison):

`ThreadPool.Instance.ChangeThreadCount(4);`

For example in the Playground.cs constructor:

```
    public Playground()
    {
        world = new World(64_000, 64_000, 32_000);
        world.NullBody.Tag = new RigidBodyTag(true);
        drawBox = DrawBox;

        ThreadPool.Instance.ChangeThreadCount(4);
    }
```

Maybe Jitter spawns too many threads for your processor. But since the single-threading performance is already bad I doubt that this fixes anything.

**Reply by MrScautHD** [reply]
Improves the FPS, but is still very bad

**Reply by MrScautHD** [reply]
With Multithreading enabled the performance get worst.

**Comment by MrScautHD** [comment]
So it looks like something is wrong with Multithreading. but i wanna use this lib it looks very well for me!

**Reply by notgiven688** [reply]
Sorry for the late reply. Can you just report me the single-threaded performance of the "Demo 03 - Ancient pyramids" scene in *release mode*:
![image](https://github.com/notgiven688/jitterphysics2/assets/37874600/773e5bfa-3cc4-41b0-8d61-5318f2c5f602)

It should be well above 100 fps for your CPU. If this is the case we have indeed a problem with multithreading.

**Reply by MrScautHD** [reply]
Its about 300 FPS
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/45c7bf2b-220a-478f-b723-7616602bf82a)

**Reply by MrScautHD** [reply]
Your Raylib Test Demo just with 8000 Cubes
====

Without Multi Threading
====
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/300b02c9-aeaa-485c-b2e8-0f588e7804b5)

With Multi Threading
========
![grafik](https://github.com/notgiven688/jitterphysics2/assets/65916181/190a943d-308f-4cc8-b4da-8afcf05ceed8)

**Comment by notgiven688** [comment]
Okay, it looks like you are right. Let us focus on a single scene (Ancient pyramids).  >200 fps single threaded is a good performance; however it looks like it does not scale well when you are using multithreading; > 1000 fps is expected for this scene. I will read into the particularities of your CPU.

**Reply by MrScautHD** [reply]
Btw is there a discord server for this project?

**Reply by notgiven688** [reply]
I created one and linked it on the main page.
